### PR TITLE
indexer-common: Implement stake feasibility check

### PR DESCRIPTION
**Pull request by tilacog**
Introduces stake feasibility check for Action batches. 

Before this, batches would be rejected even if their gross value fits in the free stake final balance.
 

Tasks: 
- [x] Batch stake feasibility check
- [x] Sort actions by GRT balance

### CODE REVIEW

File: packages/indexer-common/src/indexer-management/allocations.ts [hunk 1]
```diff
@@ -1148,4 +1157,73 @@ export class AllocationManager {
...
+    logger.debug(`Validating action batch`, { size: batch.length })
+
+    // Validate stake feasibility
+    const freeStake = await this.contracts.staking.getIndexerCapacity(this.indexer)
+    const mapper = async (action: Action) => this.resolveActionDelta(action)
+    const resolvedBatch = await pMap(batch, mapper)
+    const batchDelta: BigNumber = resolvedBatch
+      .map((resolvedAction) => resolvedAction.balance)
+      .reduce((a, b) => a.add(b))
+    const newBalance = batchDelta.add(freeStake)
```
Comment (fordN) [inline packages/indexer-common/src/indexer-management/allocations.ts:1209]: I think this should be `freeStake.sub(batchDelta)` 🤔 

From what I can tell the `resolveActionDelta()` function returns a positive value if a specific action is going to be allocating more than it is unallocating; meaning more stake will be used moving forward than currently used. As such a batch of actions will have a positive batchDelta if that batch is going to be using GRT from the freeStake pool --> batchDelta should be subtracted from freeStake to get the new freeStake value post batch execution.

File: packages/indexer-common/src/indexer-management/allocations.ts [hunk 2]
```diff
@@ -1148,4 +1157,73 @@ export class AllocationManager {
...
+    if (newBalance.isNegative()) {
+      {
+        throw indexerError(
+          IndexerErrorCode.IE013, // TODO: Should we create a new error code?
```
Comment (fordN) [inline packages/indexer-common/src/indexer-management/allocations.ts:1213]: I could go either way on this.  Perhaps we try to use the existing error code, 13, and see if that is clear enough.

**Comment by hopeyen** [comment]
hey there, #529 was implemented a couple months ago on batch stake feasibility but didn't get any feedback. I'm linking it here in case it helps, and am wondering if there's anything I can improve on in the future so that features don't have to be re-implemented?

File: packages/indexer-common/src/indexer-management/allocations.ts (hunk 1 above)
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/allocations.ts]: Fixed in https://github.com/graphprotocol/indexer/pull/579/commits/360b33ef86f211c77fcf158e75f85d139b90a0f5

File: packages/indexer-common/src/indexer-management/allocations.ts [hunk 3]
```diff
@@ -90,6 +99,17 @@ export class AllocationManager {
     private transactionManager: TransactionManager,
   ) {}
 
+  async fetchAllocation(allocationID: string, context: string): Promise<Allocation> {
```
[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/allocations.ts:102]: Handled in e01966b2.
I couldn't find occurrences where the absence of an allocation specified by ID is tolerated, so no further error handling was necessary.

[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/allocations.ts]: Handled in e01966b2.
I couldn't find occurrences where the absence of an allocation specified by ID is tolerated, so no further error handling was necessary.

[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/allocations.ts]: Seems like CI found it for me
https://github.com/graphprotocol/indexer/actions/runs/4047822351/jobs/6962307172

File: packages/indexer-common/src/indexer-management/__tests__/resolvers/actions.ts [hunk 4]
```diff
@@ -1015,3 +1032,49 @@ describe('Actions', () => {
...
   })
 })
+
+describe('Allocation Manager', () => {
+  beforeAll(setup)
+  beforeEach(setupEach)
+  afterEach(teardownEach)
+  afterAll(teardownAll)
+
+  describe('', () => {
```
[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/__tests__/resolvers/actions.ts]: Thanks for catching that. At first I thought I would need a dedicated nested test block for using the effects of upstream setup functions, but that's not the case.

Fixed in c44fe554

File: packages/indexer-common/src/indexer-management/__tests__/resolvers/actions.ts [hunk 5]
```diff
@@ -1015,3 +1032,49 @@ describe('Actions', () => {
...
+    const reallocateAction = {
+      ...invalidReallocateAction,
+      amount: 10_000,
+      id: 'reallocate action test id',
+      allocationID,
+    }
+
+    const actions = [queuedAllocateAction, unallocateAction, reallocateAction] as Action[]
+
+    test('resolveActionDelta', async () => {
```
[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-common/src/indexer-management/__tests__/resolvers/actions.ts]: Thanks for the suggestion! 

Addressed in c44fe554

**Review by fordN — APPROVED** [review]
👍

**Pull request merged**